config-model/src/main/java/com/yahoo/vespa/model/container/ApplicationContainerCluster.java

```diff
@@ -184,6 +194,60 @@ public void addServlet(Servlet servlet) {
      */
     public Optional<Integer> getMemoryPercentage() { return Optional.ofNullable(memoryPercentage); }
 
+    /*
+      Create list of endpoints, these will be consumed later by the LBservicesProducer
+     */
+    private void createEndpointList(DeployState deployState) {
+
+        if( deployState.getProperties().applicationId().instance().isTester()) return;
+        List<ApplicationClusterEndpoint> endpoints = new ArrayList<>();
+        // Add zone local endpoints using zone dns suffixes, tenant, application and cluster id.
+        // For now support both L7 and L4 routing
+
+        List<String> hosts = getContainers().stream()
+                .map(AbstractService::getHostName)
+                .collect(Collectors.toList());
+        for(String suffix : deployState.getProperties().zoneDnsSuffixes()) {
+            // L4
+            ApplicationClusterEndpoint.DnsName l4Name = ApplicationClusterEndpoint.DnsName.sharedL4NameFrom(
+                    ClusterSpec.Id.from(getName()),
+                    deployState.getProperties().applicationId(),
+                    suffix);
+            endpoints.add(ApplicationClusterEndpoint.builder()
+                                  .zoneScope()
+                                  .sharedL4Routing()
+                                  .dnsName(l4Name)
+                                  .hosts(hosts)
+                                  .build());
+
+            // L7
+            ApplicationClusterEndpoint.DnsName l7Name = ApplicationClusterEndpoint.DnsName.sharedNameFrom(
+                    ClusterSpec.Id.from(getName()),
+                    deployState.getProperties().applicationId(),
+                    suffix);
+            endpoints.add(ApplicationClusterEndpoint.builder()
+                                  .zoneScope()
+                                  .sharedRouting()
+                                  .dnsName(l7Name)
+                                  .hosts(hosts)
+                                  .build());
+        }
+
+        // Then get all endpoints provided by controller. Can be created with L4 routing only
+        Set<ContainerEndpoint> endpointsFromController = deployState.getEndpoints();
```

Comment:
This will also include application-level endpoints now (if declared), but we haven't modeled that on the config server side yet.